src/cli/rustup_mode.rs

```diff
@@ -55,17 +55,46 @@ where
 pub fn main() -> Result<utils::ExitCode> {
     self_update::cleanup_self_updater()?;
 
+    use clap::ErrorKind::*;
     let matches = match cli().get_matches_from_safe(process().args_os()) {
         Ok(matches) => Ok(matches),
-        Err(e)
-            if e.kind == clap::ErrorKind::HelpDisplayed
-                || e.kind == clap::ErrorKind::VersionDisplayed =>
-        {
-            writeln!(process().stdout().lock(), "{}", e.message)?;
+        Err(clap::Error {
+            kind: HelpDisplayed,
+            message,
+            ..
+        }) => {
+            writeln!(process().stdout().lock(), "{}", message)?;
+            return Ok(utils::ExitCode(0));
+        }
+        Err(clap::Error {
+            kind: VersionDisplayed,
+            message,
+            ..
+        }) => {
+            writeln!(process().stdout().lock(), "{}", message)?;
+            info!("This is the version for the rustup toolchain manager, not the rustc compiler.");
+
+            fn rustc_version() -> std::result::Result<String, Box<dyn std::error::Error>> {
+                let cmd = Command::new("rustc").arg("--version").output()?;
+                if cmd.status.success() {
+                    Ok(String::from_utf8_lossy(&cmd.stdout).trim().into())
+                } else {
+                    Err(String::from_utf8_lossy(&cmd.stderr).into())
+                }
+            }
```

Comment:
You don't want to use this.  Instead you want to find the default toolchain for the CWD and use `toolchain.rustc_version()`

Look at `Cfg::find_or_install_override_toolchain_or_default` for that.

Reply:
I thought about trying to find this function, but was confused how to do it without building up the whole chain. Will look again.

Just curious: Is there any downside to doing this aside from potential performance? (I'm unsure where an additional 0.1s for `--version` would matter.) Arguably, calling `rustc` as I do would also work for broken rustup installs.

Reply:
The downside is exactly as you spotted basically, also `rustup +nightly --version` will work if you look up the Cfg properly :D